Feature/quantum jobs readme (#13622)
Azure/azure-sdk-for-js: sdk/quantum/quantum-jobs/README.md

## Azure Quantum Jobs client library for JavaScript

This package contains an isomorphic SDK for QuantumJobClient.

> Azure Quantum is a Microsoft Azure service that you can use to run quantum computing programs or solve optimization problems in the cloud. Using the Azure Quantum tools and SDKs, you can create quantum programs and run them against different quantum simulators and machines. You can use the Azure.Quantum.Jobs client library to:

Reviewer: ```suggestion
Azure Quantum is a Microsoft Azure service that you can use to run quantum computing programs or solve optimization problems in the cloud. Using the Azure Quantum tools and SDKs, you can create quantum programs and run them against different quantum simulators and machines. You can use the `@azure/quantum-jobs` client library to:
```

- Create, enumerate, and cancel quantum jobs
- Enumerate provider status and quotas

>   [Source code][source] | [API reference documentation](https://docs.microsoft.com/qsharp/api/) | [Product documentation](https://docs.microsoft.com/azure/quantum/)

Reviewer: ```suggestion
  [Source code][https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/quantum/quantum-jobs] | [API reference documentation](https://docs.microsoft.com/qsharp/api/) | [Product documentation](https://docs.microsoft.com/azure/quantum/)
```

>   [Source code][source] | [API reference documentation](https://docs.microsoft.com/qsharp/api/) | [Product documentation](https://docs.microsoft.com/azure/quantum/)

Reviewer 2: We should add a link to Samples here as well

```suggestion
  [Source code][source] | [API reference documentation](https://docs.microsoft.com/qsharp/api/) | [Product documentation](https://docs.microsoft.com/azure/quantum/) | [Samples](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/quantum/quantum-jobs/samples)
```

## Getting started

This section includes everything a developer needs to install and create their first client connection _very quickly_.

### Install the package

Install the Azure Quantum Jobs client library for Javascript with `npm`:

```bash
npm install @azure/quantum-jobs
```

### Prerequisites

- Node.js version 8.x.x or higher
- [Azure subscription](https://azure.microsoft.com/free/)
- [Azure Quantum Workspace][workspaces]

### Authenticate the client

> To authenticate with the service, you can use [DefaultAzureCredential](https://docs.microsoft.com/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet). This will try different authentication mechanisms based on the environment (e.g. Environment Variables, ManagedIdentity, CachedTokens) and finally it will fallback to InteractiveBrowserCredential.

Reviewer: ```suggestion
To authenticate with the service, you can use [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/identity/identity#defaultazurecredential) from the `@azure/identity` library. This will try different authentication mechanisms based on the environment (e.g. Environment Variables, ManagedIdentity, CachedTokens) and finally, it will fallback to InteractiveBrowserCredential.
```

> To authenticate with the service, you can use [DefaultAzureCredential](https://docs.microsoft.com/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet). This will try different authentication mechanisms based on the environment (e.g. Environment Variables, ManagedIdentity, CachedTokens) and finally it will fallback to InteractiveBrowserCredential.

Reviewer: > InteractiveBrowserCredential

I'm sure it would work, please just do sanity testing before the release with [bundling docs](https://github.com/Azure/azure-sdk-for-js/blob/master/documentation/Bundling.md) since we are stating it here.

And please consider adding a section for the browser use case such as... https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/storage/storage-blob/README.md#javascript-bundle

> The client also allows the user to override the above behavior by passing their own [TokenCredential](https://docs.microsoft.com/dotnet/api/azure.core.tokencredential?view=azure-dotnet).

Reviewer: ```suggestion
The client also allows the user to override the above behavior by passing their own implementations of the [TokenCredential](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-auth/src/tokenCredential.ts).
```

`TokenCredential` is the default Authentication mechanism used by Azure SDKs.

## Key concepts

`QuantumJobClient` is the root class to be used to authenticate, and create, enumerate, and cancel jobs.

`JobDetails` contains all the properties of a job.

`ProviderStatus` contains status information for a provider.

`QuantumJobQuota` contains quota properties.

## Examples

### Create the client

Create an instance of the QuantumJobClient by passing in these parameters:

- [Subscription Id][subscriptions] - looks like XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX and can be found in your list of subscriptions on azure
- [Resource Group Name][resource-groups] - a container that holds related resources for an Azure solution
- [Workspace Name][workspaces] - a collection of assets associated with running quantum or optimization applications
- [Location][location] - choose the best data center by geographical region
- [Storage Container Name][blob-storage] - your blob storage
- [Credential][credentials] - used to authenticate

```Javascript Snippet
    const credential = new DefaultAzureCredential();

>     // Create a QuantumJobClient

Reviewer 3: const subscriptionId = process.env.SUBSCRIPTION_ID || "";
    const resourceGroupName = process.env.RESOURCE_GROUP_NAME || "";
    const workspaceName = process.env.WORKSPACE_NAME || "";
    const storageContainerName = process.env.STORAGE_CONTAINER_NAME || "";
    const location = process.env.LOCATION || "";; //"your_location";
    const endpoint = "https://" + location + ".quantum.azure.com";

Author: I'll update the samples in the next round, following the samples templates with TS and JS, and adding things like this.
Thanks for pointing it out.

    const subscriptionId = "your_subscription_id";
    const resourceGroupName = "your_resource_group_name";
    const workspaceName = "your_quantum_workspace_name";
    const storageContainerName = "mycontainer";
    const location = "westus"; //"your_location";
    const endpoint = "https://" + location + ".quantum.azure.com";

    const quantumJobClient = new QuantumJobClient(
      credential,
      subscriptionId,
      resourceGroupName,
      workspaceName,
      {
        endpoint: endpoint,

>         credentialScopes: "https://quantum.microsoft.com/.default"

Reviewer: If this scope is a constant, why not make it internal to the SDK. Why ask users to pass it?

Reviewer 3: I believe it will be defined in the SDK as part of the custom layer (in the next round)

Author: Ideally, the scope should have already been added in the generated code by autorest since we are passing 
```
add-credentials: true
credential-scopes: ["https://quantum.microsoft.com/.default"]
```

So it may be an autorest bug that I'll follow-up next week, or as Sarangan said,  add it in the custom layer.

      }
    );
```

### Get Container SAS URI

> Create a storage container where to put your data.

Reviewer: ```suggestion
Create a storage container to put your data.
```
?

```Javascript Snippet
    // Get container Uri with SAS key
    const containerUri = (
      await quantumJobClient.storage.sasUri({
        containerName: storageContainerName
      })
    ).sasUri;

    // Create container if not exists
    const containerClient = new ContainerClient(containerUri);

>     containerClient.createIfNotExists();

Reviewer: ```suggestion
    await containerClient.createIfNotExists();
```
?

    };
    const createdJob = await quantumJobClient.jobs.create(jobId, createJobDetails);
```

### Get Job

`GetJob` retrieves a specific job by its id.

```Javascript Snippet
    // Get the job that we've just created based on its jobId
    const myJob = await quantumJobClient.jobs.get(jobId);
```

### Get Jobs

To enumerate all the jobs in the workspace, use the `jobs.list` method.

```Javascript Snippet
    let jobListResult = await quantumJobClient.jobs.list();
    let listOfJobs = await jobListResult.next();
    while (!listOfJobs.done) {
      let job = listOfJobs.value;
      console.log(`  ${job.name}`);
      listOfJobs = await jobListResult.next();
    }
```

## Next steps

- Visit our [Product documentation](https://docs.microsoft.com/azure/quantum/) to learn more about Azure Quantum.

## Contributing

See the [CONTRIBUTING.md][contributing] for details on building,
testing, and contributing to this library.

This project welcomes contributions and suggestions. Most contributions require
you to agree to a Contributor License Agreement (CLA) declaring that you have
the right to, and actually do, grant us the rights to use your contribution. For

> details, visit [cla.microsoft.com](cla.microsoft.com).

Reviewer: ```suggestion
details, visit [cla.microsoft.com](https://cla.opensource.microsoft.com/).
```
?
